src/test/java/teammates/test/driver/EmailAccount.java

```diff
@@ -26,73 +34,204 @@ private EmailAccount() {
     }
 
     /**
-     * Retrieve registration key sent to Gmail inbox. After retrieving, marks
-     * the email as read.
+     * Retrieve registration key sent to Gmail inbox. After retrieving, marks the email as read.
      *
      * @return registration key (null if cannot be found).
      */
-    public static String getRegistrationKeyFromGmail(String username, String password, String courseName, String courseId)
+    public static String getRegistrationKeyFromGmail(String username, String courseName, String courseId)
             throws IOException, MessagingException {
-        Folder inbox = getGmailInbox(username, password);
-        Message[] messages = getMessages(inbox);
-
-        // Loop over up to 5 unread messages at the top of the inbox
-        int maxEmailsToCheck = Math.min(messages.length, 5);
-        for (int i = messages.length - 1; i >= messages.length - maxEmailsToCheck; i--) {
-            Message message = messages[i];
-
-            if (isRegistrationEmail(message, courseName, courseId)) {
-                String body = getEmailBody(message);
-                message.setFlag(Flags.Flag.SEEN, true);
-                inbox.close(true);
+
+        Gmail service = new GmailServiceMaker(username).makeGmailService();
+
+        ListMessagesResponse listMessagesResponse;
+        while (true) {
+            try {
+                // Get last 5 emails received by the user as there may be other emails received. However, this may fail
+                // unexpectedly if there are 5 additional emails received excluding the one from TEAMMATES.
+                listMessagesResponse = service.users().messages().list(username).setMaxResults(5L)
+                        .setQ("is:UNREAD").execute();
+                break;
+            } catch (GoogleJsonResponseException e) {
+                if (e.getDetails().getCode() == HttpStatusCodes.STATUS_CODE_FORBIDDEN) {
```

Comment:
But wouldn't it mean that if it's 403 we would have to invert the condition? Not sure if I should do this.